src/hivpy/sexual_behaviour.py

```diff
@@ -130,3 +144,143 @@ def update_sex_groups(self, population: pd.DataFrame):
+        elif(date2000 < date < date2010):
+            self.rred_population = yearly_change_90s**5
+        elif(date2010 < date < date2021):
+            dt = (date - date2010) / datetime.timedelta(days=365.25)
+            self.rred_population = yearly_change_90s**5 * yearly_change_10s**dt
+        elif(date2021 < date):
+            self.rred_population = yearly_change_90s**5 * yearly_change_10s**11
+
+    def init_rred_diagnosis(self, population):
+        population["rred_diagnosis"] = 1
+
+    def update_rred_diagnosis(self, population, date):
+        HIV_idx_new = selector(population, HIV_status=(operator.eq, True),
+                               HIV_Diagnosis_Date=(operator.ge, date-self.rred_diagnosis_period))
+        population.loc[HIV_idx_new, "rred_diagnosis"] = self.rred_diagnosis
+        HIV_idx_old = selector(population, HIV_status=(operator.eq, True),
+                               HIV_Diagnosis_Date=(operator.lt, date-self.rred_diagnosis_period))
+        population.loc[HIV_idx_old, "rred_diagnosis"] = np.sqrt(self.rred_diagnosis)
+
+    def init_rred_balance(self, population):
+        """Initialise risk reduction factor for balancing sex ratios"""
+        population["rred_balance"] = 1.0
+
+    def update_rred_balance(self, population):
+        """Update balance of new partners for consistency between sexes.
+           Integral discrepancies have been replaced with fractional discrepancy."""
+        # We first need the difference of new partners between men and women
+        men = population["sex"] == SexType.Male
+        women = population["sex"] == SexType.Female
+        mens_partners = sum(population.loc[men, "num_partners"])
+        womens_partners = sum(population.loc[women, "num_partners"])
+        partner_discrepancy = (mens_partners - womens_partners) / len(population)
+        if partner_discrepancy >= 0.1:
+            rred_balance = 0.1
+        elif partner_discrepancy >= 0.03:
+            rred_balance = 0.7
+        elif partner_discrepancy >= 0.005:
+            rred_balance = 0.7
+        elif partner_discrepancy >= 0.004:
+            rred_balance = 0.75
+        elif partner_discrepancy >= 0.003:
+            rred_balance = 0.8
+        elif partner_discrepancy >= 0.002:
+            rred_balance = 0.9
+        elif partner_discrepancy >= 0.001:
+            rred_balance = 0.97
+        elif partner_discrepancy >= -0.001:
+            rred_balance = 1
+        elif partner_discrepancy >= -0.002:
+            rred_balance = 1/0.97
+        elif partner_discrepancy >= -0.003:
+            rred_balance = 1/0.9
+        elif partner_discrepancy >= -0.004:
+            rred_balance = 1/0.8
+        elif partner_discrepancy >= -0.005:
+            rred_balance = 1/0.75
+        elif partner_discrepancy >= -0.03:
+            rred_balance = 1/0.7
+        elif partner_discrepancy >= -0.1:
+            rred_balance = 1/0.7
+        else:
+            rred_balance = 10
```

Comment:
Can we encode the numbers only once, for positive discrepancies, and then check which way we have to adjust? Perhaps in a separate function that gives the ratio when passed a positive discrepancy. So we either use `self.balance_factor(discrepancy)` or `1 / self.balance_factor(-discrepancy)` (and the reciprocal for the other sex)

Looking at the numbers again, maybe it's more complicated than `-discrepancy`, but hopefully the idea is clear!